sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClientBuilder.java

```diff
@@ -352,4 +360,77 @@ private ProxyConfiguration getDefaultProxyConfiguration(Configuration configurat
 
         return new ProxyConfiguration(authentication, proxy, username, password);
     }
+
+    /**
+     * This property must be set for building an {@link EventProcessorAsyncClient}.
+     *
+     * The consumer group name from which the {@link EventProcessorAsyncClient} should consume
+     * the events from
+     * @param consumerGroupName The consumer group name
+     * @return The updated {@link EventHubClientBuilder} object.
+     */
+    public EventHubClientBuilder consumerGroupName(String consumerGroupName) {
+        this.consumerGroupName = consumerGroupName;
+        return this;
+    }
+
+    /**
+     * This property can be optionally set when building an {@link EventProcessorAsyncClient}
+     *
+     * This event position will be used if event position for a partition is not available.
+     * If this property is not set and if checkpoint for a partition doesn't exist, {@link EventPosition#earliest()}
+     * will be used as the initial event position to start consuming events.
+     *
+     * @param initialEventPosition The initial event position
+     * @return The updated {@link EventHubClientBuilder} object.
+     */
+    public EventHubClientBuilder initialEventPosition(EventPosition initialEventPosition) {
+        this.initialEventPosition = initialEventPosition;
+        return this;
+    }
+
+    /**
+     * This property must be set or a class implementing {@link PartitionManager} should be available in classpath
+     * when building an {@link EventProcessorAsyncClient}
+     *
+     * @param partitionManager Implementation of {@link PartitionManager}
+     * @return The updated {@link EventHubClientBuilder} object.
+     */
+    public EventHubClientBuilder partitionManager(PartitionManager partitionManager) {
+        // If this is not set, look for classes implementing PartitionManager interface
```

Comment:
I am fine with this being discussed. My basic proposition was based on an assumption that in most cases the user will likely have on their classpath (i.e. with their application) a single partition manager. This would be on the classpath because the user opted into a specific one as part of their Maven configuration. Having opted into this explicitly, we could take this as their choice being made and discover / wire this in for them automatically. If we discover that there are zero or more than one Partition Manager's on the classpath, we can choose to throw an exception at construction time.

This is similar in design to how we support pluggable HTTP policies, and soon pluggable HTTP clients.

Reply:
I feel like we're in some very Java-specific areas here, and I don't want to push against common practice, known patterns, and language idioms, so I'm going to nod and smile at most of that since I just don't have the context or experience to really discuss intelligently.

For my own edification, what happens if I have multiple candidates in the classpath that would satisfy the requirement?   Do we then force explicit selection or just opt for first-in-wins-because-please-don't-define-multiples?   Would I be able to have an implementation with multiple candidates that I can configure injection for at runtime?   (in other words, is this more of a controllable dependency injection for Java similar to how other ecosystems use IOC containers?)

Reply:
This is where we can define the rules that we want. A set of rules we might define are:

1. If there is no partition manager (PM) on the classpath, and if the user does not explicitly provide one (somehow?!), then we throw an exception.
2. If there is no PM on the classpath, and the user does specify one, we use it.
3. If there is exactly one PM found on the classpath, and if the user does not specify one, we use the one we found.
4. If there is exactly one PM found on the classpath, and the user specifies a different one, we use the one the user specified.
5. If there is more than one PM found on the classpath, and the user does not specify one, we can either choose to take the first one we find, or to throw an exception (I would opt for an exception).
6. If there is more than one PM found on the classpath, and the user does specify one, we use the one the user specified.

Reply:
Thanks for the insight.  For what its worth, that all sounds lucid and well-reasoned to me, from an outsider's perspective.